File: tools/elgato_stream_deck_plugins/adi_ableton_vst_controller/ableton/remote_script/AdiVST/ws_server.py
# -*- coding: utf-8 -*-
"""
Minimal, dependency-free WebSocket server (RFC 6455) for the AdiVST Remote Script.

Why hand-rolled: an Ableton MIDI Remote Script runs inside Live's bundled
CPython with no pip, so we cannot `import websockets`. This implements just
enough of RFC 6455 to talk to a browser/CEF client: the HTTP upgrade handshake,
masked client text frames in, unmasked server text frames out, and ping/pong/
close control frames.

Threading model: the server runs on its own daemon thread with a `select` loop.
It NEVER touches the Live API. Inbound text messages are handed to `on_message`
(called on the server thread) — the Remote Script must marshal anything that
touches Live onto Live's main thread. Outbound messages are queued with
`broadcast()` from any thread and flushed by the loop within ~`SELECT_TIMEOUT`.
"""

import base64
import collections
import errno
import hashlib
import select
import socket
import struct
import threading

_GUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"
SELECT_TIMEOUT = 0.04          # s — bounds outbound latency and lets the thread exit
MAX_FRAME = 4 * 1024 * 1024    # reject any single frame / runaway recv buffer above this
MAX_OUT_BUF = 1 << 20          # drop a client whose unsent buffer exceeds this (slow reader)


def _accept_key(client_key):
    sha = hashlib.sha1((client_key + _GUID).encode("utf-8")).digest()
    return base64.b64encode(sha).decode("ascii")


class _Client(object):
    """Per-connection state: handshake buffer, frame parser, outbound bytes."""

    def __init__(self, sock, addr):
        self.sock = sock
        self.addr = addr
        self.handshook = False
        self.in_buf = bytearray()
        self.out_buf = bytearray()
        self.frag_op = None
        self.frag_data = bytearray()
        self.alive = True

    def fileno(self):
        return self.sock.fileno()

    def queue(self, data):
        # backpressure: drop a stalled/slow client rather than grow memory unbounded
        if len(self.out_buf) + len(data) > MAX_OUT_BUF:
            self.alive = False
            return
        self.out_buf.extend(data)


class WSServer(threading.Thread):
    def __init__(self, host="127.0.0.1", port=9006, on_message=None, on_connect=None, log=None):
        threading.Thread.__init__(self)
        self.daemon = True
        self.host = host
        self.port = port
        self.on_message = on_message or (lambda text, client: None)
        self.on_connect = on_connect or (lambda client: None)
        self.log = log or (lambda *a: None)
        self._lsock = None
        self._clients = []
        self._stop = threading.Event()
        self._outbox = collections.deque()  # (client_or_None, text) — thread-safe append/pop

    # ---------------------------------------------------------------- public
    def broadcast(self, text, client=None):
        """Queue a text message to all clients (or one). Safe from any thread."""
        self._outbox.append((client, text))

    def client_count(self):
        return len([c for c in self._clients if c.handshook])

    def stop(self):
        self._stop.set()
        try:
            if self._lsock:
                # nudge the select loop awake
                self._lsock.close()
        except Exception:
            pass

    # ------------------------------------------------------------- thread run
    def run(self):
        try:
            self._lsock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self._lsock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            self._lsock.bind((self.host, self.port))
            self._lsock.listen(8)
            self._lsock.setblocking(False)
            self.log("AdiVST WS listening on %s:%d" % (self.host, self.port))
        except Exception as e:
            self.log("AdiVST WS bind failed: %s" % e)
            return

        while not self._stop.is_set():
            self._flush_outbox()
            rlist = [self._lsock] + [c.sock for c in self._clients]
            wlist = [c.sock for c in self._clients if c.out_buf]
            try:
                r, w, _ = select.select(rlist, wlist, [], SELECT_TIMEOUT)
            except (OSError, ValueError):
                # a socket was closed mid-select; prune and retry
                self._prune()
                continue

            for s in r:
                if s is self._lsock:
                    self._accept()
                else:
                    self._read(self._client_for(s))
            for s in w:
                self._send(self._client_for(s))

            self._prune()

        self._shutdown()

    # ----------------------------------------------------------- internals
    def _client_for(self, sock):
        for c in self._clients:
            if c.sock is sock:
                return c
        return None

    def _accept(self):
        try:
            sock, addr = self._lsock.accept()
            sock.setblocking(False)
            self._clients.append(_Client(sock, addr))
        except Exception:
            pass

    def _read(self, c):
        if c is None:
            return
        try:
            data = c.sock.recv(65536)
        except socket.error as e:
            if e.args and e.args[0] in (errno.EWOULDBLOCK, errno.EAGAIN):
                return
            c.alive = False
            return
        if not data:
            c.alive = False
            return
        c.in_buf.extend(data)
        if len(c.in_buf) > MAX_FRAME + 4096:   # runaway / wedged buffer guard
            c.alive = False
            return
        if not c.handshook:
            self._try_handshake(c)
        if c.handshook:
            self._parse_frames(c)

    def _send(self, c):
        if c is None or not c.out_buf:
            return
        try:
            sent = c.sock.send(c.out_buf)
            del c.out_buf[:sent]
        except socket.error as e:
            if e.args and e.args[0] in (errno.EWOULDBLOCK, errno.EAGAIN):
                return
            c.alive = False

    def _flush_outbox(self):
        while True:
            try:
                target, text = self._outbox.popleft()
            except IndexError:
                break
            frame = self._encode_text(text)
            if target is not None:
                if target.handshook:
                    target.queue(frame)
            else:
                for c in self._clients:
                    if c.handshook:
                        c.queue(frame)

    def _prune(self):
        dead = [c for c in self._clients if not c.alive]
        for c in dead:
            try:
                c.sock.close()
            except Exception:
                pass
            self._clients.remove(c)

    def _shutdown(self):
        for c in self._clients:
            try:
                c.sock.close()
            except Exception:
                pass
        self._clients = []
        try:
            if self._lsock:
                self._lsock.close()
        except Exception:
            pass

    # --------------------------------------------------------- HTTP handshake
    def _try_handshake(self, c):
        end = c.in_buf.find(b"\r\n\r\n")
        if end < 0:
            if len(c.in_buf) > 16384:
                c.alive = False
            return
        header = bytes(c.in_buf[:end]).decode("latin-1", "ignore")
        del c.in_buf[:end + 4]
        key = None
        for line in header.split("\r\n")[1:]:
            if ":" in line:
                k, v = line.split(":", 1)
                if k.strip().lower() == "sec-websocket-key":
                    key = v.strip()
        if not key:
            c.alive = False
            return
        resp = (
            "HTTP/1.1 101 Switching Protocols\r\n"
            "Upgrade: websocket\r\n"
            "Connection: Upgrade\r\n"
            "Sec-WebSocket-Accept: %s\r\n\r\n" % _accept_key(key)
        )
        c.queue(resp.encode("latin-1"))
        c.handshook = True
        try:
            self.on_connect(c)
        except Exception as e:
            self.log("on_connect error: %s" % e)

    # -------------------------------------------------------------- framing
    def _parse_frames(self, c):
        buf = c.in_buf
        while True:
            if len(buf) < 2:
                return
            b0, b1 = buf[0], buf[1]
            fin = b0 & 0x80
            opcode = b0 & 0x0F
            masked = b1 & 0x80
            length = b1 & 0x7F
            idx = 2
            if length == 126:
                if len(buf) < 4:
                    return
                length = struct.unpack(">H", bytes(buf[2:4]))[0]
                idx = 4
            elif length == 127:
                if len(buf) < 10:
                    return
                length = struct.unpack(">Q", bytes(buf[2:10]))[0]
                idx = 10
            # RFC 6455: the 64-bit length's high bit must be 0; also cap the size
            # so a bogus/huge announced length can't wedge the parser or balloon memory.
            if (length >> 63) or length > MAX_FRAME:
                c.alive = False
                return
            if masked:
                if len(buf) < idx + 4:
                    return
                mask = buf[idx:idx + 4]
                idx += 4
            if len(buf) < idx + length:
                return
            payload = bytearray(buf[idx:idx + length])
            if masked:
                for i in range(length):
                    payload[i] ^= mask[i & 3]
            del buf[:idx + length]

            # RFC 6455 §5.5: control frames must not be fragmented and must be <= 125 bytes
            if opcode in (0x8, 0x9, 0xA) and (not fin or length > 125):
                c.queue(self._encode_frame(0x8, struct.pack(">H", 1002)))
                c.alive = False
                return

            if opcode == 0x8:  # close
                code = bytes(payload[:2]) if len(payload) >= 2 else b""
                c.queue(self._encode_frame(0x8, code))   # echo peer's status code
                c.alive = False
                return
            elif opcode == 0x9:  # ping -> pong
                c.queue(self._encode_frame(0xA, bytes(payload)))
            elif opcode == 0xA:  # pong
                pass
            elif opcode == 0x0:  # continuation
                c.frag_data.extend(payload)
                if fin:
                    self._dispatch(c, c.frag_op, c.frag_data)
                    c.frag_op = None
                    c.frag_data = bytearray()
            elif opcode in (0x1, 0x2):  # text / binary
                if fin:
                    self._dispatch(c, opcode, payload)
                else:
                    c.frag_op = opcode
                    c.frag_data = bytearray(payload)

    def _dispatch(self, c, opcode, data):
        if opcode != 0x1:  # we only consume text
            return
        try:
            text = bytes(data).decode("utf-8")
        except Exception:
            return
        try:
            self.on_message(text, c)
        except Exception as e:
            self.log("on_message error: %s" % e)

    def _encode_text(self, text):
        return self._encode_frame(0x1, text.encode("utf-8"))

    def _encode_frame(self, opcode, payload):
        n = len(payload)
        header = bytearray()
        header.append(0x80 | opcode)  # FIN + opcode
        if n < 126:
            header.append(n)
        elif n < 65536:
            header.append(126)
            header.extend(struct.pack(">H", n))
        else:
            header.append(127)
            header.extend(struct.pack(">Q", n))
        return bytes(header) + payload

File: tools/elgato_stream_deck_plugins/adi_studio_os/service/ws-server.js
// Minimal RFC 6455 WebSocket server, localhost only, zero dependencies.
//
// Studio OS carries small JSON control messages (MIDI notes, CC, keystroke
// tokens, short replies), so a full library would be dead weight and one more
// thing to vendor, audit and keep current. This handles exactly what the IPC
// contract needs: the handshake, masked client text frames including fragmented
// and 16/64-bit-length ones, ping/pong, and a clean close.
//
// It refuses any connection that is not from the loopback interface. The socket
// can drive MIDI and synthesise keystrokes, so it must never be reachable off
// this machine even if the bind address is ever changed by mistake.
import http from "node:http";
import crypto from "node:crypto";

const GUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";

const OP = { CONT: 0x0, TEXT: 0x1, BIN: 0x2, CLOSE: 0x8, PING: 0x9, PONG: 0xa };
const HEARTBEAT_MS = 15000;  // ping cadence; a client that misses one is dropped
const MAX_MESSAGE = 1 << 20; // 1 MiB — far above any real payload; a bigger
                             // frame means something is wrong, so we close.

export class WsServer {
  constructor({ port = 9011, host = "127.0.0.1", logger = console } = {}) {
    this.port = port;
    this.host = host;
    this.log = logger;
    this.clients = new Set();
    this.onMessage = () => {};
    this.onConnect = () => {};
    this.onDisconnect = () => {};
    this.server = null;
  }

  listen() {
    this.server = http.createServer((req, res) => {
      // A plain GET is useful for "is the service alive?" from a shell.
      res.writeHead(200, { "content-type": "application/json" });
      res.end(JSON.stringify({ service: "studioos", clients: this.clients.size }));
    });

    this.server.on("upgrade", (req, socket, head) => this._upgrade(req, socket, head));
    /* V64 — A LISTEN FAILURE MUST EXIT, NOT LIMP ON. This used to log and keep
       running: measured, after EADDRINUSE the process stayed alive forever with
       no listener, so every key on the board was a silent no-op and KeepAlive
       could not rescue it because it never exited. Now the supervisor gets its
       chance. Non-listen errors still only log — those are per-connection. */
    this.server.on("error", (e) => {
      this.log.error?.(`ws listen failed: ${e.message}`);
      if (e && (e.code === "EADDRINUSE" || e.code === "EACCES")) {
        this.log.error?.(`port ${this.port} is unusable — exiting so the supervisor can retry`);
        process.exit(1);
      }
    });
    this.server.listen(this.port, this.host, () => {
      this.log.info?.(`studioos service listening on ws://${this.host}:${this.port}`);
    });

    /* Heartbeat. Observed on the real install: after a few Stream Deck app
       restarts the service reported 3 clients when only 1 existed. A CEF page
       that goes away without a clean close leaves a half-open TCP connection
       that never fires 'close', so the socket count only ever grows.

       That is not cosmetic. "Silence every sounding note when the LAST client
       disconnects" is the guarantee that stops a crash mid-nudge from leaving a
       note stuck on in rekordbox — and a phantom client means the last real one
       leaving never looks like the last one. */
    this.heartbeat = setInterval(() => {
      for (const c of [...this.clients]) {
        if (!c.awaitingPong) { c.awaitingPong = true; c.ping(); continue; }
        this.log.warn?.("dropping a client that missed its pong (half-open socket)");
        c.terminate();
      }
    }, HEARTBEAT_MS);
    this.heartbeat.unref?.();   // never hold the process open on its own
    return this;
  }

  _upgrade(req, socket, head) {
    const addr = socket.remoteAddress || "";
    const local = addr === "127.0.0.1" || addr === "::1" || addr === "::ffff:127.0.0.1";
    const key = req.headers["sec-websocket-key"];
    if (!local || !key) {
      socket.end("HTTP/1.1 403 Forbidden\r\n\r\n");
      return;
    }

    const accept = crypto.createHash("sha1").update(key + GUID).digest("base64");
    socket.write(
      "HTTP/1.1 101 Switching Protocols\r\n" +
      "Upgrade: websocket\r\n" +
      "Connection: Upgrade\r\n" +
      `Sec-WebSocket-Accept: ${accept}\r\n\r\n`
    );
    socket.setNoDelay(true); // control messages must not wait on Nagle

    const client = new WsClient(socket, this);
    this.clients.add(client);
    this.onConnect(client);
    if (head && head.length) client._feed(head);

    socket.on("data", (chunk) => client._feed(chunk));
    socket.on("error", () => client._gone());
    socket.on("close", () => client._gone());
    /* V64 — 'end' IS THE EVENT THAT ACTUALLY FIRES, and its absence was the whole
       bug. http.Server builds its sockets with allowHalfOpen: true, so when a CEF
       page vanishes the server socket gets 'end' and then stays writable FOREVER
       — 'close' is never emitted. Measured on the app's own bundled node 20.20.0:
       'end' true, 'close' false at +300ms, +1s and +2s.

       So the only reaper was the 15 s heartbeat, and it needs TWO cycles because
       cycle 1 only arms awaitingPong. Thirty seconds of a phantom client — during
       which index.js's `clients.size === 0` never becomes true, so the REAL client
       leaving never triggers midi.panicAll() and a held note can stay sounding.
       That is the one guarantee this service's own header promises.

       `terminate()` rather than `_gone()`: the peer is gone, so there is nothing
       to hand a close handshake to, and destroy() is what frees the fd. */
    socket.on("end", () => client.terminate());
  }

  broadcast(obj) {
    const s = JSON.stringify(obj);
    for (const c of this.clients) c.sendRaw(s);
  }

  close() {
    clearInterval(this.heartbeat);
    for (const c of [...this.clients]) c.close();
    this.server?.close();
  }
}

class WsClient {
  constructor(socket, server) {
    this.socket = socket;
    this.server = server;
    this.buf = Buffer.alloc(0);
    this.fragments = [];   // continuation frames of the current message
    this.fragOp = null;
    this.alive = true;
    this.awaitingPong = false;
  }

  ping() {
    if (!this.alive) return;
    try { this.socket.write(encode(OP.PING, Buffer.alloc(0))); }
    catch { this._gone(); }
  }

  // Hard drop for a socket that stopped answering — close() waits on a TCP
  // handshake the peer will never complete.
  terminate() {
    if (!this.alive) return;
    try { this.socket.destroy(); } catch { /* already torn down */ }
    this._gone();
  }

  send(obj) { this.sendRaw(JSON.stringify(obj)); }

  sendRaw(text) {
    if (!this.alive) return false;
    try { this.socket.write(encode(OP.TEXT, Buffer.from(text, "utf8"))); return true; }
    catch { this._gone(); return false; }
  }

  close(code = 1000) {
    if (!this.alive) return;
    try {
      const b = Buffer.alloc(2);
      b.writeUInt16BE(code, 0);
      this.socket.write(encode(OP.CLOSE, b));
      this.socket.end();
    } catch { /* already gone */ }
    this._gone();
  }

  /* V64 — _gone() USED TO LEAVE THE SOCKET OPEN with all three listeners still
     attached. Since the heartbeat only walks `server.clients`, once _gone() had
     run nothing could ever reap that socket again — and `_feed` kept buffering on
     it. Measured: 1 MiB retained per continuation frame, no ceiling, on a client
     the service had already reported as disconnected.

     Destroying here is safe for the graceful path too: close() writes its CLOSE
     frame and calls end() BEFORE calling us, so the bytes are already queued. */
  _gone() {
    if (!this.alive) return;
    this.alive = false;
    this.server.clients.delete(this);
    try { this.socket.destroy(); } catch { /* already torn down */ }
    this.buf = Buffer.alloc(0);
    this.fragments = [];
    this.fragOp = null;
    this.server.onDisconnect(this);
  }

  _feed(chunk) {
    /* V64 — a dead client must not still dispatch. Without this, verbs arriving
       in the same tick as the teardown ran AFTER onDisconnect: measured, two of
       them. A midi.noteOn there is armed after the final panicAll() and, because
       the client is already out of the set, nothing will ever panic again — a
       permanently stuck note. */
    if (!this.alive) return;
    this.buf = this.buf.length ? Buffer.concat([this.buf, chunk]) : chunk;
    for (;;) {
      const frame = decode(this.buf);
      if (!frame) return;                       // need more bytes
      if (frame.error) { this.close(1002); return; }
      this.buf = this.buf.subarray(frame.size);

      this.awaitingPong = false;   // any frame proves the peer is alive
      if (frame.op === OP.CLOSE) { this.close(1000); return; }
      if (frame.op === OP.PING) {
        try { this.socket.write(encode(OP.PONG, frame.payload)); } catch { this._gone(); }
        continue;
      }
      if (frame.op === OP.PONG) { this.awaitingPong = false; continue; }

      // Text / continuation. Reassemble before parsing so a split JSON payload
      // is never handed to JSON.parse half-formed.
      if (frame.op === OP.CONT) {
        if (this.fragOp === null) { this.close(1002); return; }
        this.fragments.push(frame.payload);
      } else {
        if (this.fragOp !== null) { this.close(1002); return; }
        this.fragOp = frame.op;
        this.fragments = [frame.payload];
      }

      const total = this.fragments.reduce((n, b) => n + b.length, 0);
      if (total > MAX_MESSAGE) { this.close(1009); return; }
      if (!frame.fin) continue;

      const payload = this.fragments.length === 1 ? this.fragments[0] : Buffer.concat(this.fragments);
      const wasText = this.fragOp === OP.TEXT;
      this.fragments = []; this.fragOp = null;
      if (!wasText) continue;

      let msg;
      try { msg = JSON.parse(payload.toString("utf8")); } catch { continue; }
      this.server.onMessage(msg, this);
    }
  }
}

// ---------------------------------------------------------------- framing
function encode(op, payload) {
  const len = payload.length;
  let header;
  if (len < 126) {
    header = Buffer.alloc(2);
    header[1] = len;
  } else if (len < 65536) {
    header = Buffer.alloc(4);
    header[1] = 126;
    header.writeUInt16BE(len, 2);
  } else {
    header = Buffer.alloc(10);
    header[1] = 127;
    header.writeBigUInt64BE(BigInt(len), 2);
  }
  header[0] = 0x80 | op;   // FIN + opcode; the server never fragments
  return Buffer.concat([header, payload]);
}

// Returns null when more bytes are needed, { error } on a protocol violation,
// otherwise { fin, op, payload, size }.
function decode(buf) {
  if (buf.length < 2) return null;
  const fin = (buf[0] & 0x80) !== 0;
  const rsv = buf[0] & 0x70;
  const op = buf[0] & 0x0f;
  const masked = (buf[1] & 0x80) !== 0;
  let len = buf[1] & 0x7f;
  let off = 2;

  if (rsv) return { error: true };            // no extensions negotiated
  if (!masked) return { error: true };        // clients MUST mask (RFC 6455 §5.1)

  if (len === 126) {
    if (buf.length < off + 2) return null;
    len = buf.readUInt16BE(off); off += 2;
  } else if (len === 127) {
    if (buf.length < off + 8) return null;
    const big = buf.readBigUInt64BE(off); off += 8;
    if (big > BigInt(MAX_MESSAGE)) return { error: true };
    len = Number(big);
  }

  if (buf.length < off + 4 + len) return null;
  const mask = buf.subarray(off, off + 4); off += 4;
  const payload = Buffer.allocUnsafe(len);
  for (let i = 0; i < len; i++) payload[i] = buf[off + i] ^ mask[i & 3];

  return { fin, op, payload, size: off + len };
}
